File: python/collatz-conjecture/collatz_conjecture.py
from functools import wraps


def check_if_number(func):
    @wraps(func)
    def wrapper(number):
        return is_number(number) and func(number)
    return wrapper

    
def is_number(number):
    if number <= 0:
        raise ValueError("Error: expected a positive number!")
    else:
        return number


@check_if_number
def is_even(number):
    return number // 2


@check_if_number
def is_odd(number):
    return number * 3 + 1


@check_if_number
def steps(number):
    count_steps = 0
    while number > 1:
      if number % 2 == 0:
        number = is_even(number)
      else:
        number = is_odd(number)
      count_steps += 1
    return count_steps

File: go/collatz-conjecture/collatz_conjecture.go
// Package collatzconjecture solves the Collatz Conjecture challenge
package collatzconjecture

import "fmt"

// IsEven divides n by 2 to get n / 2, if n is even.
func IsEven(n int) int {
	return n / 2
}
// IsOdd multiplies n by 3 and adds 1 to get 3n + 1, if n is odd.
func IsOdd(n int) int {
	return n * 3 + 1
}
// CollatzConjecture given a number n,
// return the number of steps required to reach 1.
func CollatzConjecture(number int) (int, error) {
	countSteps := 0
	if number <= 0 {
		err := fmt.Errorf("invalid value: %q", number)
		return 0, err
	}
	for number > 1 {
		if number % 2 == 0 {
			number = IsEven(number)
		} else {
			number = IsOdd(number)
		}
		countSteps++
	}
	return countSteps, nil
}
